internal/common/mqtt/mqttconfig.go

```diff
@@ -49,14 +49,10 @@ type Client struct {
 	protocol string
 }
 
-//Message is used to wrap the app id and payload into one and publish to broker
-type Message struct {
-	AppID   string
-	Payload string
-}
-
 var (
-	clientData map[string]*Client
+	publishData     map[string]*Client
+	subscribeData   map[string]string    //topic-->published data
+	subscribeClient map[string][]*Client //topic-->multiple clients
```

Comment:
What about not considering `appid`? I think it's very inefficient if a lot of different clients sends messages to same topic.
I also think we can merge these variables into one.

Reply:
I hve followed the design where any number of clients can publish message on any topic and so appid is not used. But only those clients who have subscribed to the topic should receive the message. 
Keeping the vairable separate is to decouple the information, as the subscribeClient can hold the data for client and ill be used to delete the client wwhen the client is disconnceted so client object is required.Only appid will not help.
The subscribeData is used to store the topics mapped to the messages and this has no relation with client object. So to decouple the information they are kept separate.

Reply:
Assume that there are 3 applications that want to publish messages about `home` topic. In this case, current code create 3 `Client` variables for the same topic and broker will send same message to the 3 `Client` variable.

> I hve followed the design where any number of clients can publish message on any topic and so appid is not used. But only those clients who have subscribed to the topic should receive the message. 

I think this case we can control `Client` by `topic` instead of `clientID` and `Client` has a list to manage `clientID`.

> Keeping the vairable separate is to decouple the information, as the subscribeClient can hold the data for client and ill be used to delete the client wwhen the client is disconnceted so client object is required.Only appid will not help. The subscribeData is used to store the topics mapped to the messages and this has no relation with client object. So to decouple the information they are kept separate.

I disagree with the information should be decoupled. If the `Client` is managed by `topic`, it can hold the message instead of `subscribeData`. Also the `Client` can be removed if there is no item in the `clientID` list.

Reply:
No, client cant be managed by topic, Mosquitto requires unique client id and topic can be repetative and so. If the same client is publishing data on same topic or diff topic, he will not be added 3 times, since that client is already present and so same connection object shall be used and only publishing shall happen. Since the data is stored topic wise, if the client published 3 message on home only the latest shall be present and not three times

Reply:
What about combining `topic` and `device iID` to create a `client ID` for broker?
If we look at the `edge-orchestration` as a client, it's very inefficient to create multiple objects on the same topic.
For example, Edge Orchestration is subscribing the topic 'home' with 'app1', 'app2', and 'app3'
If `app` publishes a message, Edge Orchestration receives same 3 messages like log below,
```
INFO[2022-03-24T07:52:59Z]mqttconnection.go:136 Publish [MQTTConnectionMgr] Publishing the data to cloud
INFO[2022-03-24T07:52:59Z]mqttconnection.go:156 func1 [MQTTConnectionMgr] Topic home registered.. {"AppID":"app","Payload":"test"} is the payload
INFO[2022-03-24T07:52:59Z]mqttconnection.go:156 func1 [MQTTConnectionMgr] Topic home registered.. {"AppID":"app","Payload":"test"} is the payload
INFO[2022-03-24T07:52:59Z]mqttconnection.go:156 func1 [MQTTConnectionMgr] Topic home registered.. {"AppID":"app","Payload":"test"} is the payload
```

Reply:
In terms of broker, we can use a combination of 'topic' and 'device id'.
In terms of edge orchestration, we use a combination of 'topic' and 'host' for hash.

Reply:
> In terms of broker, we can use a combination of 'topic' and 'device id'

No, Because Topic,deviceId can be repeatative as deviceID will be same for all clients and multiple clients canhave same topics. For example HomeEdge DeviceId is 12345 same for any client, now client1 wants to register/publish on topic kitchen and client 2 also on kitchen So both clients will have kitchen,12345 so uniquness not present

Reply:
> > In terms of broker, we can use a combination of 'topic' and 'device id'
> 
> No, Because Topic,deviceId can be repeatative as deviceID will be same for all clients and multiple clients canhave same topics. For example HomeEdge DeviceId is 12345 same for any client, now client1 wants to register/publish on topic kitchen and client 2 also on kitchen So both clients will have kitchen,12345 so uniquness not present

Yes. That's my opinion. In terms of broker, we don't need to distinguish connections by client id.

Reply:
So, u mean all the third party apps sending connection request, shall create only one connection, with client id deviceId,topic
So going forward, when the client calls disconnects which connection shall be terminated by homeedge?
Also then logically there will be one connection between broker and all client, And all clients shall be virtually added to hashmap on publish/subscribe method and removed on disconnect call.
Correct?
If that is the case then only deviceId is sufficient ,since broker needs only one variable for uniqueness and not combination f values.

Reply:
> So, u mean all the third party apps sending connection request, shall create only one connection, with client id deviceId,topic So going forward, when the client calls disconnects which connection shall be terminated by homeedge?

Assume that edge-orchestration can have a requester list for host `broker.com` and topic `home`. 
There were 3 client id (app1, app2, app3) requested. If `app1` calls disconnection, Then edge-orchestration deletes `app1` from the requester list. If there is no client id in the requester list, we can terminate the connection between broker and edge-orchestration.  

> Also then logically there will be one connection between broker and all client, And all clients shall be virtually added to hashmap on publish/subscribe method and removed on disconnect call. Correct? If that is the case then only deviceId is sufficient ,since broker needs only one variable for uniqueness and not combination f values.

All client connects to edge-orchestration, not broker. Also I don't understand why we should subscribe same topic to same broker with multiple client ID because the data from broker with a topic is same. Subscribed data is not different by client ID.